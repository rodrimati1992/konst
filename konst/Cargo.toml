[package]
name = "konst"
version = "0.3.5"
authors = ["rodrimati1992 <rodrimatt1985@gmail.com>"]
rust-version = "1.65.0"
edition = "2021"
license = "Zlib"
description = "Const equivalents of std functions, compile-time comparison, and parsing"
documentation = "https://docs.rs/konst/"
readme="../README.md"
keywords = ["no-std", "const_fn", "parsing"]
categories = ["no-std", "parsing"]
repository = "https://github.com/rodrimati1992/konst/"
include = [
    "Cargo.toml", 
    "src/**/*.rs", 
    "../README.md",
    "LICENSE-ZLIB.md", 
]


[dependencies.konst_proc_macros]
version = "=0.3.0"
path = "../konst_proc_macros"
optional = true

[dependencies.konst_kernel]
version = "=0.3.5"
path = "../konst_kernel"
features = ["rust_1_64", "__for_konst"]

[dev-dependencies.rand]
version = "0.8.4"
default-features = false
features = ["small_rng"]

# dev-dependencies can't be optional
[dependencies.trybuild]
version = "1.0"
optional = true

[dependencies.typewit]
version = "1.1"
features = ["rust_1_61"]

[dependencies.const_panic]
version = "0.2.6"
features = ["rust_1_64"]
default-features = false

[features]
default = ["cmp", "iter", "parsing_proc"]


# Enables extra checks for debug
debug = ["konst_kernel/debug"]
cmp = []
iter = ["konst_kernel/iter"]
parsing_proc = ["parsing", "konst_proc_macros"]
parsing = []
rust_latest_stable = []
mut_refs = ["rust_latest_stable", "konst_kernel/mut_refs"]
nightly_mut_refs = ["mut_refs", "konst_kernel/nightly_mut_refs"]

alloc = ["konst_kernel/alloc"]

# Enables ui tests, which are intended for the latest stable release.
# This is not enabled in CI, because it's way too sensitive to 
# changes in how rustc formats errors
__ui = ["trybuild", "rust_latest_stable"]
__cp_derive = ["const_panic/derive", "const_panic/non_basic"]

docsrs = ["konst_kernel/docsrs"]



[package.metadata.docs.rs]
features = ["docsrs", "alloc", "cmp", "parsing_proc", "rust_latest_stable", "nightly_mut_refs"]

