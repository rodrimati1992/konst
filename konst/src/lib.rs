//! Const equivalents of std functions and const parsing.
//!
//! # Features
//!
//! This crate provides:
//!
//! - Const fn equivalents of standard library functions, methods, and operators.
//!
//! - [`destructure`]/[`if_let_Some`]/[`while_let_Some`]
//! macros to allow destructuring types in const without getting "cannot drop in const" errors.
//!
//! - Compile-time parsing through the [`Parser`] type, and [`parser_method`] macro.
//!
//! # Examples
//!
//! ### Parsing an enum
//!
//! This example demonstrates how you can parse a simple enum from an environment variable,
//! at compile-time.
//!
//! ```rust
//! use konst::{eq_str, option, result};
//! use konst::const_panic::{self, PanicFmt};
//!
//! use std::fmt::{self, Display};
//!
//!
//! const CHOICE: &str = option::unwrap_or!(option_env!("chosen-direction"), "forward");
//!
//! const DIRECTION: Direction = result::unwrap!(Direction::try_parse(CHOICE));
//!
//! fn main() {
//!     match DIRECTION {
//!         Direction::Forward => assert_eq!(CHOICE, "forward"),
//!         Direction::Backward => assert_eq!(CHOICE, "backward"),
//!         Direction::Left => assert_eq!(CHOICE, "left"),
//!         Direction::Right => assert_eq!(CHOICE, "right"),
//!     }
//! }
//!
//!
//! #[derive(Debug, PartialEq)]
//! enum Direction {
//!     Forward,
//!     Backward,
//!     Left,
//!     Right,
//! }
//!
//! impl Direction {
//!     const fn try_parse(input: &str) -> Result<Self, ParseDirectionError> {
//!         // As of Rust 1.89.0, string patterns don't work in const contexts
//!         match () {
//!             _ if eq_str(input, "forward") => Ok(Direction::Forward),
//!             _ if eq_str(input, "backward") => Ok(Direction::Backward),
//!             _ if eq_str(input, "left") => Ok(Direction::Left),
//!             _ if eq_str(input, "right") => Ok(Direction::Right),
//!             _ => Err(ParseDirectionError),
//!         }
//!     }
//! }
//!
//! // `PanicFmt` derives the `PanicFmt` trait for debug-printing in `result::unwrap`.
//! // To use the `PanicFmt` derive you need to enable the "const_panic_derive" feature.
//! #[derive(Debug, PartialEq, PanicFmt)]
//! pub struct ParseDirectionError;
//!
//! impl Display for ParseDirectionError {
//!     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
//!         f.write_str("Failed to parse a Direction")
//!     }
//! }
//! ```
//!
//! ### Parsing CSV
//!
//! This example demonstrates how CSV can be parsed into integers.
//!
//! This example requires the `"iter"` feature (enabled by default).
//!
#![cfg_attr(feature = "iter", doc = "```rust")]
#![cfg_attr(not(feature = "iter"), doc = "```ignore")]
//! use konst::{iter, result, string};
//!
//!
//! assert_eq!(PARSED, [3, 8, 13, 21, 34]);
//!
//!
//! const CSV: &str = "3, 8, 13, 21, 34";
//!
//! static PARSED: [u64; 5] = iter::collect_const!(u64 =>
//!     string::split(CSV, ","),
//!         map(str::trim_ascii),
//!         map(|s| result::unwrap!(u64::from_str_radix(s, 10))),
//! );
//! ```
//!
//! ### Parsing a struct
//!
//! This example demonstrates how a key-value pair format can be parsed into a struct.
//!
//! This requires the `"iter"` and `"parsing_proc"` features (enabled by default).
//!
#![cfg_attr(all(feature = "iter", feature = "parsing_proc"), doc = "```rust")]
#![cfg_attr(
    not(all(feature = "iter", feature = "parsing_proc")),
    doc = "```ignore"
)]
//! use konst::{result, try_};
//! use konst::parsing::{Parser, ParseError, parser_method};
//!
//!
//! fn main(){
//!     assert_eq!(
//!         PARSED,
//!         Struct{
//!             name: "bob smith",
//!             amount: 1000,
//!             repeating: Shape::Circle,
//!             colors: [Color::Red, Color::Blue, Color::Green, Color::Blue],
//!         }
//!     );
//! }
//!
//!
//! const PARSED: Struct = {
//!     // You can also parse strings from environment variables, or from an `include_str!(....)`
//!     let input = "\
//!         colors = red, blue, green, blue
//!         amount = 1000
//!         repeating = circle
//!         name = bob smith
//!     ";
//!     
//!     result::unwrap!(parse_struct(&mut Parser::new(input)))
//! };
//!
//! #[derive(Debug, Clone, PartialEq, Eq)]
//! pub struct Struct<'a> {
//!     pub name: &'a str,
//!     pub amount: usize,
//!     pub repeating: Shape,
//!     pub colors: [Color; 4],
//! }
//!
//! #[derive(Debug, Clone, PartialEq, Eq)]
//! pub enum Shape {
//!     Circle,
//!     Square,
//!     Line,
//! }
//!
//! #[derive(Debug, Copy, Clone, PartialEq, Eq)]
//! pub enum Color {
//!     Red,
//!     Blue,
//!     Green,
//! }
//!
//! pub const fn parse_struct<'p>(parser: &mut Parser<'p>) -> Result<Struct<'p>, ParseError<'p>> {
//!     let mut name = "<none>";
//!     let mut amount = 0;
//!     let mut repeating = Shape::Circle;
//!     let mut colors = [Color::Red; 4];
//!     
//!     parser.trim_end();
//!     if !parser.is_empty() {
//!         loop {
//!             // keeps a copy of the parser at this position to parse the field name
//!             let mut prev_parser = parser.trim_start().copy();
//!
//!             // skip past the `=` to parse the field value
//!             try_!(parser.find_skip('=')).trim_start();
//!
//!             parser_method!{prev_parser, strip_prefix;
//!                 "name" => name = try_!(parser.split_keep('\n')),
//!                 "amount" => amount = try_!(parser.parse_usize()),
//!                 "repeating" => repeating = try_!(parse_shape(parser)),
//!                 "colors" => colors = try_!(parse_colors(parser)),
//!                 _ => {
//!                     let err = &"could not parse Struct field name";
//!                     return Err(prev_parser.to_other_error(err));
//!                 }
//!             }
//!
//!             if parser.is_empty() {
//!                 break
//!             }
//!             try_!(parser.strip_prefix("\n"));
//!         }
//!     }
//!
//!     Ok(Struct{name, amount, repeating, colors})
//! }
//!
//! pub const fn parse_shape<'p>(parser: &mut Parser<'p>) -> Result<Shape, ParseError<'p>> {
//!     let shape = parser_method!{parser, strip_prefix;
//!         "circle" => Shape::Circle,
//!         "square" => Shape::Square,
//!         "line" => Shape::Line,
//!         _ => return Err(parser.to_other_error(&"could not parse Shape"))
//!     };
//!     Ok(shape)
//! }
//!
//! pub const fn parse_colors<'p, const LEN: usize>(
//!     parser: &mut Parser<'p>,
//! ) -> Result<[Color; LEN], ParseError<'p>> {
//!     let mut colors = konst::array::ArrayBuilder::of_copy();
//!
//!     while !colors.is_full() {
//!         colors.push(try_!(parse_color(parser.trim_start())));
//!         
//!         // returns an error if there aren't enough comma-separated colors
//!         if let Err(e) = parser.strip_prefix(",") && !colors.is_full() {
//!             return Err(e);
//!         }
//!     }
//!
//!     Ok(colors.build())
//! }
//!
//! pub const fn parse_color<'p>(parser: &mut Parser<'p>) -> Result<Color, ParseError<'p>> {
//!     let color = parser_method!{parser, strip_prefix;
//!         "red" => Color::Red,
//!         "blue" => Color::Blue,
//!         "green" => Color::Green,
//!         _ => return Err(parser.to_other_error(&"could not parse Color"))
//!     };
//!     Ok(color)
//! }
//!
//! ```
//!
//! # Cargo features
//!
//! These are the features of this crate:
//!
//! - `"iter"`(enabled by default):
//! Enables all iteration-related items that take/return iterators,
//!
//! - `"cmp"`(enabled by default):
//! Enables all comparison-related items,
//! the string equality and ordering comparison functions don't require this feature.
//!
//! - `"parsing_proc"`(enabled by default):
//! Enables the `"parsing"` feature, compiles the `konst_proc_macros` dependency,
//! and enables the [`parser_method`] macro.
//! You can use this feature instead of `"parsing"` if the slightly longer
//! compile times aren't a problem.
//!
//! - `"parsing"`(enabled by default):
//! Enables the [`parsing`] module for parsing from strings,
//! and the `primitive::parse_*` functions.
//!
//! - `"const_panic_derive"`(disabled by default):
//! Enables the "derive" feature of the `const_panic` public dependency.
//!
//! - `"alloc"`(disabled by default):
//! Enables items that use types from the [`alloc`] crate.
//!
//! ### Rust release related
//!
//! None of thse features are enabled by default.
//!
//! - `"rust_latest_stable"`: enables the latest `"rust_1_*"` feature.
//! Only recommendable if you can update the Rust compiler every stable release.
//!
//! # No-std support
//!
//! `konst` is `#![no_std]`, it can be used anywhere Rust can be used.
//!
//! # Minimum Supported Rust Version
//!
//! `konst` requires Rust 1.89.0.
//!
//! Features that require newer versions of Rust, or the nightly compiler,
//! need to be explicitly enabled with crate features.
//!
//!
//!
//! [`alloc`]: https://doc.rust-lang.org/alloc/
//! [`const_eq`]: crate::cmp::const_eq
//! [`const_eq_for`]: crate::cmp::const_eq_for
//! [`const_cmp`]: crate::cmp::const_cmp
//! [`const_cmp_for`]: crate::cmp::const_cmp_for
//! [`parsing`]: crate::parsing
//! [`primitive`]: crate::primitive
//! [`parser_method`]: crate::parsing::parser_method
//! [`Parser`]: crate::parsing::Parser
//! [`Parser::parse_u128`]: crate::parsing::Parser#method.parse_u128
//! [`destructure`]: crate::destructure
//! [`if_let_Some`]: crate::if_let_Some
//! [`while_let_Some`]: crate::while_let_Some
//!
#![deny(missing_docs)]
#![deny(unused_results)]
#![allow(rustdoc::redundant_explicit_links)]
// clippy's opinionated BS
#![allow(clippy::needless_doctest_main)]
#![allow(clippy::init_numbered_fields)]
#![allow(clippy::redundant_pattern_matching)]
#![allow(clippy::unused_unit)]
////////////
#![forbid(clippy::missing_const_for_fn)]
#![cfg_attr(feature = "docsrs", feature(doc_cfg))]
#![no_std]

#[cfg(feature = "alloc")]
extern crate alloc;

#[macro_use]
#[doc(hidden)]
pub mod macros;

mod docs;

#[doc(hidden)]
pub mod __for_cmp_impls;

#[doc(hidden)]
pub mod __utils;

#[cfg(feature = "alloc")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "alloc")))]
pub mod alloc_type;

pub mod array;

pub mod chr;

#[cfg(feature = "cmp")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
pub mod cmp;

#[cfg(feature = "iter")]
#[macro_use]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "iter")))]
pub mod iter;

pub mod drop_flavor;

pub mod ffi;

pub mod polymorphism;

pub mod primitive;

pub mod option;

pub mod result;

pub mod range;

pub mod maybe_uninit;

pub mod manually_drop;

pub mod nonzero;

pub mod other;

#[cfg(feature = "parsing")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "parsing")))]
pub mod parsing;

pub mod ptr;

mod utils;

#[cfg(feature = "parsing")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "parsing")))]
pub use crate::parsing::Parser;

#[cfg(feature = "parsing_proc")]
#[doc(hidden)]
pub use konst_proc_macros::{__priv_bstr_end, __priv_bstr_start};

pub mod slice;

pub mod string;

pub use ::const_panic;

pub use crate::string::{cmp_str, eq_str};

#[cfg(feature = "cmp")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
pub use crate::string::{cmp_option_str, eq_option_str};

#[cfg(all(doctest, feature = "iter", feature = "parsing_proc"))]
#[doc = include_str!("../../README.md")]
pub struct ReadmeTest;

#[doc(hidden)]
pub mod __ {
    pub use core::{
        assert,
        cmp::Ordering::{self, Equal, Greater, Less},
        compile_error, concat,
        marker::PhantomData,
        matches,
        mem::{self, ManuallyDrop, MaybeUninit, replace},
        ops::Range,
        option::Option::{self, None, Some},
        panic,
        primitive::{bool, str, u8, usize},
        ptr,
        result::Result::{self, Err, Ok},
        stringify,
    };

    pub const fn forget<T>(x: T) {
        core::mem::forget(x)
    }

    pub mod v {
        pub use core::{
            option::Option::Some,
            result::Result::{Err, Ok},
        };
    }

    pub use crate::__for_cmp_impls::U8Ordering;

    #[cfg(feature = "cmp")]
    pub use crate::cmp::{CmpWrapper, ConstCmp, IsAConstCmp, IsNotStdKind, IsStdKind};

    #[cfg(feature = "iter")]
    pub use crate::iter::collect_const::CollectorCmd;

    pub use crate::array::__array_macros_2::unit_array;

    pub use const_panic::concat_panic;

    pub use typewit::MakeTypeWitness;
}
