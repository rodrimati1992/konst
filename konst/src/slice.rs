//! `const fn` equivalents of slice methods.
//!
//! # Removed in 0.3.0
//!
//! These functions were removed in 0.3.0 because there is an equivalent
//! const fn in the standard library:
//!
//! - `first`
//! - `last`
//! - `split_first`
//! - `split_last`
//!
//!
//!

/// `const fn`s for comparing slices for equality and ordering.
#[cfg(feature = "cmp")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
pub mod cmp;

mod bytes_pattern;
mod slice_as_chunks;
mod slice_concatenation;
mod slice_const_methods;

#[cfg(feature = "iter")]
mod slice_iter_methods;

pub use bytes_pattern::BytesPattern;

pub(crate) use bytes_pattern::PatternNorm;

pub use self::slice_as_chunks::*;
pub use self::slice_concatenation::*;
pub use self::slice_const_methods::*;

#[cfg(feature = "iter")]
pub use slice_iter_methods::*;

__declare_slice_cmp_fns! {
    import_path = "konst",

    (
        ///
        ///  # Example
        ///
        ,
        /// ```rust
        /// use konst::slice::eq_bytes;
        ///
        /// const FOO: &[u8] = b"foo";
        /// const BAR: &[u8] = b"fooooo";
        /// const BAZ: &[u8] = b"bar";
        ///
        ///
        /// const FOO_EQ_FOO: bool = eq_bytes(FOO, FOO);
        /// assert!( FOO_EQ_FOO );
        ///
        /// const FOO_EQ_BAR: bool = eq_bytes(FOO, BAR);
        /// assert!( !FOO_EQ_BAR );
        ///
        /// const FOO_EQ_BAZ: bool = eq_bytes(FOO, BAZ);
        /// assert!( !FOO_EQ_BAZ );
        ///
        /// ```
        ///
        ,
        /// ```rust
        /// use konst::slice::cmp_bytes;
        ///
        /// use std::cmp::Ordering;
        ///
        /// const FOO: &[u8] = b"foo";
        /// const BAR: &[u8] = b"fooooo";
        /// const BAZ: &[u8] = b"bar";
        ///
        ///
        /// const FOO_CMP_FOO: Ordering = cmp_bytes(FOO, FOO);
        /// assert_eq!(FOO_CMP_FOO, Ordering::Equal);
        ///
        /// const FOO_CMP_BAR: Ordering = cmp_bytes(FOO, BAR);
        /// assert_eq!(FOO_CMP_BAR, Ordering::Less);
        ///
        /// const FOO_CMP_BAZ: Ordering = cmp_bytes(FOO, BAZ);
        /// assert_eq!(FOO_CMP_BAZ, Ordering::Greater);
        ///
        /// ```
        ///
        ,
        u8,
        eq_bytes,
        cmp_bytes,
    )
}

__declare_fns_with_docs! {
    (Option<&'a [u8]>, (eq_option_bytes, cmp_option_bytes))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a,]
        params(l, r)
        eq_comparison = eq_bytes(l, r),
        cmp_comparison = cmp_bytes(l, r),
        parameter_copyability = copy,
    ),
}

/// The error produced by trying to convert from
/// `&[T]` to `&[T; N]`, or from `&mut [T]` to `&mut [T; N]`.
#[doc(inline)]
pub use konst_kernel::slice::slice_for_konst::TryIntoArrayError;

/// Tries to convert from `&[T]` to `&[T; N]`.
///
/// Returns an `Err(TryIntoArrayError{..})` when the slice doesn't match the expected length.
///
/// # Example
///
/// ```rust
/// use konst::{
///     slice::{TryIntoArrayError, try_into_array},
///     result,
///     unwrap_ctx,
/// };
///
///
/// const fn arr_5() -> Option<&'static [u64; 5]> {
///     let slice: &[u64] = &[1, 10, 100, 1000, 10000];
///
///     // Passing the length explicitly to the function
///     result::ok!(try_into_array::<_, 5>(slice))
/// }
///
/// assert_eq!(arr_5(), Some(&[1, 10, 100, 1000, 10000]));
///
///
/// const fn err() -> Result<&'static [u64; 5], TryIntoArrayError> {
///     let slice: &[u64] = &[];
///
///     // Letting the function infer the length of the array,
///     try_into_array(slice)
/// }
///
/// assert!(err().is_err());
///
///
/// const fn arr_3() -> &'static [u64; 3] {
///     let slice: &[u64] = &[3, 5, 8];
///
///     let array = unwrap_ctx!(try_into_array(slice));
///     
///     // You can destructure the array into its elements like this
///     let [a, b, c] = *array;
///     
///     array
/// }
///
/// assert_eq!(arr_3(), &[3, 5, 8]);
///
/// ```
///
/// [`try_into_array`]: ./macro.try_into_array.html
#[doc(inline)]
pub use konst_kernel::slice::slice_for_konst::try_into_array_func as try_into_array;

/// Tries to convert from `&mut [T]` to `&mut [T; N]`.
///
/// Returns an `Err(TryIntoArrayError{..})` when the slice doesn't match the expected length.
///
/// # Example
///
/// ```rust
/// # #![feature(const_mut_refs)]
/// use konst::{slice, unwrap_ctx};
///
/// const fn mut_array_from<const LEN: usize>(slice: &mut [u8], from: usize) -> &mut [u8; LEN] {
///     let sliced = slice::slice_range_mut(slice, from, from + LEN);
///     unwrap_ctx!(slice::try_into_array_mut(sliced))
/// }
///
/// # fn main() {
///
/// let slice = &mut [3, 5, 8, 13, 21, 34, 55, 89, 144, 233];
///
/// let foo: &mut [u8; 2] = mut_array_from(slice, 0);
/// assert_eq!(foo, &mut [3, 5]);
///
/// let bar: &mut [u8; 3] = mut_array_from(slice, 2);
/// assert_eq!(bar, &mut [8, 13, 21]);
///
/// let baz: &mut [u8; 4] = mut_array_from(slice, 4);
/// assert_eq!(baz, &mut [21, 34, 55, 89]);
///
/// # }
/// ```
///
#[cfg(feature = "mut_refs")]
#[cfg_attr(
    feature = "docsrs",
    doc(cfg(any(feature = "mut_refs", feature = "nightly_mut_refs")))
)]
#[doc(inline)]
pub use konst_kernel::slice::slice_for_konst::try_into_array_mut_func as try_into_array_mut;
