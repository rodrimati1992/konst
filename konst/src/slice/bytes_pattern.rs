use crate::{
    chr,
    polymorphism::{HasTypeWitness, MakeTypeWitness, TypeEq, TypeWitnessTypeArg},
};

/// A byte slice pattern.
///
/// Types that implement this trait can be used to search into a byte slice.
///
/// This trait can only be implemented in the `konst` crate.
///
pub trait BytesPattern<const N: usize>: HasTypeWitness<BytesPatternInput<N, Self>> {
    #[doc(hidden)]
    type __Normalized: ?Sized;
}

#[non_exhaustive]
pub enum BytesPatternInput<const N: usize, P: ?Sized + BytesPattern<N>> {
    Str(TypeEq<P, str>, TypeEq<*const P::__Normalized, *const [u8]>),
    Bytes(TypeEq<P, [u8]>, TypeEq<*const P::__Normalized, *const [u8]>),
    Array(
        TypeEq<P, [u8; N]>,
        TypeEq<*const P::__Normalized, *const [u8]>,
    ),
    Char(
        TypeEq<P, char>,
        TypeEq<*const P::__Normalized, *const chr::Utf8Encoded>,
    ),
}

impl<const N: usize, Arg> TypeWitnessTypeArg for BytesPatternInput<N, Arg>
where
    Arg: ?Sized + BytesPattern<N>,
{
    type Arg = Arg;
}

impl MakeTypeWitness for BytesPatternInput<0, str> {
    const MAKE: Self = BytesPatternInput::Str(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for str {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl MakeTypeWitness for BytesPatternInput<0, [u8]> {
    const MAKE: Self = BytesPatternInput::Bytes(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for [u8] {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl<const N: usize> MakeTypeWitness for BytesPatternInput<N, [u8; N]> {
    const MAKE: Self = BytesPatternInput::Array(TypeEq::NEW, TypeEq::NEW);
}
impl<const N: usize> BytesPattern<N> for [u8; N] {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl MakeTypeWitness for BytesPatternInput<0, char> {
    const MAKE: Self = BytesPatternInput::Char(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for char {
    #[doc(hidden)]
    type __Normalized = chr::Utf8Encoded;
}

#[derive(Copy, Clone)]
pub(crate) enum PatternNorm<'a, const N: usize, P>
where
    P: ?Sized + BytesPattern<N>,
{
    Bytes {
        val: &'a [u8],
        ten: TypeEq<*const P::__Normalized, *const [u8]>,
    },
    Char {
        val: chr::Utf8Encoded,
        ten: TypeEq<*const P::__Normalized, *const chr::Utf8Encoded>,
    },
}

impl<'a, const N: usize, P> PatternNorm<'a, N, P>
where
    P: ?Sized + BytesPattern<N>,
{
    pub(crate) const fn new(pattern: &'a P) -> Self {
        match P::WITNESS {
            BytesPatternInput::Str(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern).as_bytes();
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Bytes(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern);
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Array(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern);
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Char(te, ten) => {
                let val: &'a char = te.in_ref().to_right(pattern);
                let val = chr::encode_utf8(*val);
                PatternNorm::Char { val, ten }
            }
        }
    }

    pub(crate) const fn as_bytes(&self) -> &[u8] {
        match self {
            PatternNorm::Bytes { val, ten } => ten.reachability_hint(val),
            PatternNorm::Char { val, ten } => ten.reachability_hint(val.as_bytes()),
        }
    }
}
