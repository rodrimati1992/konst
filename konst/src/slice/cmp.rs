#[doc(inline)]
pub use crate::slice::{
    cmp_bytes as cmp_slice_u8, cmp_option_bytes as cmp_option_slice_u8, eq_bytes as eq_slice_u8,
    eq_option_bytes as eq_option_slice_u8,
};

use core::cmp::Ordering;

__declare_slice_cmp_fns! {
    import_path = "konst",

    (,,, u16, eq_slice_u16, cmp_slice_u16,)
    (,,, u32, eq_slice_u32, cmp_slice_u32,)
    (,,, u64, eq_slice_u64, cmp_slice_u64,)
    (,,, u128, eq_slice_u128, cmp_slice_u128,)
    (,,, usize, eq_slice_usize, cmp_slice_usize,)

    (,,, i8, eq_slice_i8, cmp_slice_i8,)
    (,,, i16, eq_slice_i16, cmp_slice_i16,)
    (,,, i32, eq_slice_i32, cmp_slice_i32,)
    (,,, i64, eq_slice_i64, cmp_slice_i64,)
    (,,, i128, eq_slice_i128, cmp_slice_i128,)
    (,,, isize, eq_slice_isize, cmp_slice_isize,)

    (,,, bool, eq_slice_bool, cmp_slice_bool,)
    (,,, char, eq_slice_char, cmp_slice_char,)
}

__delegate_const_eq! {
    skip_coerce;

    /// Compares two `&[&str]` for equality.
    pub const fn eq_slice_str(ref l: &[&str], r: &[&str]) -> bool {
        crate::const_eq_for!(slice; l, r, crate::eq_str)
    }
}

__delegate_const_ord! {
    skip_coerce;

    /// Compares two `&[&str]`, returning the order of `left` relative to `right`.
    pub const fn cmp_slice_str(ref left: &[&str], right: &[&str]) -> Ordering {
        crate::const_cmp_for!(slice; left, right, crate::cmp_str)
    }
}

__delegate_const_eq! {
    skip_coerce;

    /// Compares two `&[&[u8]]` for equality.
    pub const fn eq_slice_bytes(ref l: &[&[u8]], r: &[&[u8]]) -> bool {
        crate::const_eq_for!(slice; l, r, eq_slice_u8)
    }
}

__delegate_const_ord! {
    skip_coerce;

    /// Compares two `&[&[u8]]`, returning the order of `left` relative to `right`.
    pub const fn cmp_slice_bytes(ref left: &[&[u8]], right: &[&[u8]]) -> Ordering {
        crate::const_cmp_for!(slice; left, right, cmp_slice_u8)
    }
}

__declare_fns_with_docs! {
    (Option<&'a [u16]>, (eq_option_slice_u16, cmp_option_slice_u16))
    (Option<&'a [u32]>, (eq_option_slice_u32, cmp_option_slice_u32))
    (Option<&'a [u64]>, (eq_option_slice_u64, cmp_option_slice_u64))
    (Option<&'a [u128]>, (eq_option_slice_u128, cmp_option_slice_u128))
    (Option<&'a [usize]>, (eq_option_slice_usize, cmp_option_slice_usize))
    (Option<&'a [i8]>, (eq_option_slice_i8, cmp_option_slice_i8))
    (Option<&'a [i16]>, (eq_option_slice_i16, cmp_option_slice_i16))
    (Option<&'a [i32]>, (eq_option_slice_i32, cmp_option_slice_i32))
    (Option<&'a [i64]>, (eq_option_slice_i64, cmp_option_slice_i64))
    (Option<&'a [i128]>, (eq_option_slice_i128, cmp_option_slice_i128))
    (Option<&'a [isize]>, (eq_option_slice_isize, cmp_option_slice_isize))
    (Option<&'a [bool]>, (eq_option_slice_bool, cmp_option_slice_bool))
    (Option<&'a [char]>, (eq_option_slice_char, cmp_option_slice_char))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a,]
        params(l, r)
        eq_comparison = crate::cmp::CmpWrapper(l).const_eq(r),
        cmp_comparison = crate::cmp::CmpWrapper(l).const_cmp(r),
        parameter_copyability = copy,
    ),
}

__declare_fns_with_docs! {
    (Option<&'a [&'b str]>, (eq_option_slice_str, cmp_option_slice_str))
    (Option<&'a [&'b [u8]]>, (eq_option_slice_bytes, cmp_option_slice_bytes))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a, 'b,]
        params(l, r)
        eq_comparison = crate::cmp::CmpWrapper(l).const_eq(r),
        cmp_comparison = crate::cmp::CmpWrapper(l).const_cmp(r),
        parameter_copyability = copy,
    ),
}
