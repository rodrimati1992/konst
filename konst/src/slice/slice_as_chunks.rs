use crate::slice;

/// Const equivalent of
/// [`<[T]>::as_chunks`](https://doc.rust-lang.org/std/primitive.slice.html#method.as_chunks)
///
/// # Panics
///
/// Panics if `N == 0`.
///
/// # Example
///
/// ```rust
/// use konst::slice;
///
/// let (arrs, rem) = slice::as_chunks::<_, 3>(&[2u32, 3, 5, 8, 13, 21, 34, 55]);
///
/// assert_eq!(arrs, &[[2, 3, 5], [8, 13, 21]][..]);
/// assert_eq!(rem, &[34, 55][..])
///
/// ```
#[track_caller]
pub const fn as_chunks<'a, T, const N: usize>(this: &[T]) -> (&[[T; N]], &[T]) {
    assert!(N != 0, "chunk size must be non-zero");

    let arrs_len = this.len() / N;
    let (arrs_in, rem) = slice::split_at(this, arrs_len * N);

    // SAFETY: `arrs_in` is a `&[T]` that is `arrs_len * N` long,
    // its layout is compatible with the `&[[T; N]]` that this produces.
    let arrs: &[[T; N]] =
        unsafe { core::slice::from_raw_parts(arrs_in.as_ptr() as *const [T; N], arrs_len) };

    (arrs, rem)
}

/// Const equivalent of
/// [`<[T]>::as_rchunks`](https://doc.rust-lang.org/std/primitive.slice.html#method.as_rchunks)
///
/// # Panics
///
/// Panics if `N == 0`.
///
/// # Example
///
/// ```rust
/// use konst::slice;
///
/// let (rem, arrs) = slice::as_rchunks::<_, 2>(&[2u32, 3, 5, 8, 13, 21, 34]);
///
/// assert_eq!(rem, &[2][..]);
/// assert_eq!(arrs, &[[3, 5], [8, 13], [21, 34]][..]);
/// ```
#[track_caller]
pub const fn as_rchunks<'a, T, const N: usize>(this: &[T]) -> (&[T], &[[T; N]]) {
    assert!(N != 0, "chunk size must be non-zero");

    let arrs_len = this.len() / N;
    let rem_len = this.len() % N;
    let (rem, arrs_in) = slice::split_at(this, rem_len);

    // SAFETY: `arrs_in` is a `&[T]` that is `arrs_len * N` long,
    // its layout is compatible with the `&[[T; N]]` that this produces.
    let arrs: &[[T; N]] =
        unsafe { core::slice::from_raw_parts(arrs_in.as_ptr() as *const [T; N], arrs_len) };

    (rem, arrs)
}
