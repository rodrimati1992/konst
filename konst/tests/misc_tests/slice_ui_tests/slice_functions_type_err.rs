use konst::slice;

struct NonCopy;

const _: () = {
    let arr = [const { NonCopy }; 4];

    slice::fill(&mut arr, NonCopy);
};



fn main() {}

