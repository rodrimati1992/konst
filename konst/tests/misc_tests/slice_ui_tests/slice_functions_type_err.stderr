error[E0277]: the trait bound `NonCopy: Copy` is not satisfied
 --> tests/misc_tests/slice_ui_tests/slice_functions_type_err.rs:8:5
  |
8 |     slice::fill(&mut arr, NonCopy);
  |     ^^^^^^^^^^^ the trait `Copy` is not implemented for `NonCopy`
  |
note: required by a bound in `fill`
 --> src/slice/slice_filler.rs
  |
  | pub const fn fill<T: Copy>(slice: &mut [T], val: T) {
  |                      ^^^^ required by this bound in `fill`
help: consider annotating `NonCopy` with `#[derive(Copy)]`
  |
3  + #[derive(Copy)]
4  | struct NonCopy;
   |
